pallets/attesters/src/lib.rs

```diff
@@ -517,49 +819,219 @@ pub mod pallet {
+        }
+
+        pub fn get_batch_by_message(
+            target: TargetId,
+            message: Vec<u8>,
+        ) -> Option<BatchMessage<T::BlockNumber>> {
+            match Batches::<T>::get(target) {
+                Some(batches) => batches.iter().find(|&b| b.message() == message).cloned(),
+                None => None,
+            }
+        }
+
+        pub fn get_batch_by_message_hash(
+            target: TargetId,
+            message_hash: H256,
+        ) -> Option<BatchMessage<T::BlockNumber>> {
+            match Batches::<T>::get(target) {
+                Some(batches) => batches
+                    .iter()
+                    .find(|&b| b.message_hash() == message_hash)
+                    .cloned(),
+                None => None,
+            }
+        }
+
+        pub fn get_latest_batch_to_sign(target: TargetId) -> Option<BatchMessage<T::BlockNumber>> {
+            let mut batches = Self::get_batches(target, BatchStatus::PendingAttestation);
+            batches.sort_by(|a, b| b.created.cmp(&a.created));
+            batches.first().cloned()
+        }
+
+        pub fn get_latest_batch_to_sign_hash(target: TargetId) -> Option<H256> {
+            let mut batches = Self::get_batches(target, BatchStatus::PendingAttestation);
+            batches.sort_by(|a, b| b.created.cmp(&a.created));
+            batches.iter().map(|b| b.message_hash()).next()
+        }
+
+        pub fn get_latest_batch_to_sign_message(target: TargetId) -> Option<Vec<u8>> {
+            let mut batches = Self::get_batches(target, BatchStatus::PendingAttestation);
+            batches.sort_by(|a, b| b.created.cmp(&a.created));
+            batches.iter().map(|b| b.message()).next()
+        }
+
+        fn update_sorted_nominated_attesters(
+            attester: &T::AccountId,
+            amount: BalanceOf<T>,
+        ) -> bool {
+            let mut all_indices_match = true;
+            SortedNominatedAttesters::<T>::mutate(|attesters| {
+                if let Some(index) = attesters.iter().position(|(a, _n)| a == attester) {
+                    let total_nomination = attesters[index].1 - amount;
+                    if total_nomination.is_zero() {
+                        attesters.remove(index);
+                    } else {
+                        attesters[index] = (attester.clone(), total_nomination);
+                    }
+
+                    // Sort the attesters by their nomination amount
+                    attesters.sort_by(|(_a1, n1), (_a2, n2)| n2.cmp(n1));
+
+                    // Keep only the top 32 attesters in the list
+                    attesters.truncate(32);
+                } else {
+                    log::warn!("Attester not found while updating sorted nominated attesters");
+                    all_indices_match = false
+                }
+            });
+            all_indices_match
+        }
+
+        pub fn do_nominate(
+            nominator: T::AccountId,
+            attester: T::AccountId,
```

Comment:
These should take a reference to avoid cloning all the time when we use it

Reply:
addressed [a30be0d](https://github.com/t3rn/t3rn/pull/980/commits/a30be0dec0af88cd2c5cde2f7176eddf8fa89489)

Reply:
addressed [a30be0d](https://github.com/t3rn/t3rn/pull/980/commits/a30be0dec0af88cd2c5cde2f7176eddf8fa89489)